### ISSUE TITLE
Cache not being hit for single items, that were previously requested in a list

### ISSUE THREAD

**Issue by lukasluecke** [FormidableLabs/urql-exchange-graphcache#86]
I was under the impression that the following queries should result in the cache being hit, but it just sends two requests. (I load the `ItemList` in a main screen, and load `Item` inside a subview, **after** clicking on a list item). The `__typename`s are correctly requested and match.

```graphql
query ItemList {
    items {
        edges {
            node {
                id
                name
            }
        }
    }
}
```

```graphql
query Item($id: ID!) {
    item(id: $id) {
        id
        name
    }
}
```

**Comment by lukasluecke** [comment]
@JoviDeCroock Here's the issue I'm running into right now 😅

**Comment by JoviDeCroock** [comment]
Yes this is normal, we can't assume that `items` queries that `item` since `__typename` is received from the server. This would need us to do assumptions about your data, we could implement some new config in graphCache that does the following:

```js
fromCollection: {
  item: (args, cache) => cache.getEntity(someRandomTypename, args.id) 
}
```

This is essentially a load-time/performance optimisation.

**Comment by lukasluecke** [comment]
@JoviDeCroock That's what I figured - but I still think it should be possible (or even work like that by default). Since the `node` field has the same `__typename` and `id` as the `item` field, should that not mean that they share a cache "slot"? The partial handling should be similar to what we have for other cases already, no? And at least if the fields are identical (or a superset) we could just return those.

**Comment by JoviDeCroock** [comment]
Yes, we only know the `__typename` from the server response or if we would infer it from our introspection result. Doing a partial return without introspection result is dangerous, so it would only be safe for when there's a schema present in my opinion. That would imply that we see if the __typename matches and set our result from the start with that result. Then it continues as normal. If it sees that it can partially return it will and then dispatch an additional operation to query the missing fields.

**Comment by JoviDeCroock** [comment]
Do note that the second time your list and details query WILL be linked in the cache since now it DOES know that for that entity that the details query writes to the same location as the list.id one. This means that if you would write an update/optimistic it would update both.

**Comment by lukasluecke** [comment]
@JoviDeCroock Providing a way to "match" the fields with a introspection result / schema would be great. It's not really that big of an issue right now, it would just be "nice" to have 😉

**Comment by JoviDeCroock** [comment]
I will probably try and discuss this later with some others since I'm not entirely sure how feasible it is to work on schema alone. This would require some premises, one we would need to be sure that there is an argument stating we want an id (and that should actually be the only argument), then we could go ahead and check if a record exists with that specific name and identifier.

This would then result in us instead of dispatching the query to a server, querying our own cache. This could lead to a partial result --> return data to application and fetch the other data.

Personally this is exclusive for scenario's where we have an introspected schema.

TLDR'ed this so I can point to it in conversations.

**Comment by lukasluecke** [comment]
Sounds good 👍 Thanks for the detailed response as well 🙂 

I was using Apollo in the past, which is why that was my expectation in the first place. Not sure how much sense their approach makes, I prefer stability in this case - but if you can figure out a good solution that would be awesome!

**Comment by JoviDeCroock** [comment]
Hey @lukasluecke 

For now your use case can be solved by:

```
  const cache = cacheExchange({resolvers: {
    Query: {
      todo: (_parent, args, cache) => {
       return { __typename: 'Todo', id: args.id }
      }
    },
 } });
```

We don't want to go to automatic stuff since there could be a lot of server logic that won't be triggered if we do so.

**Comment by lukasluecke** [comment]
@JoviDeCroock Thanks, I'll try that soon. I'd still appreciate a more "automatic" solution in the future, even if it's not fully automatic (e.g. you have to specify a map of Queries / Typenames - but not define identical functions for each of them), or opt-in (e.g. only for `cache-only`?).

**Comment by lukasluecke** [comment]
@JoviDeCroock Just wanted to let you know that your solution worked for my current use case 🙂

**Comment by JoviDeCroock** [comment]
All right, closing this issue :D thanks for bringing it to our attention, we have added it to the documentation. We'll work on better documentation in the coming days.
